File: backend/app/contracts/run_manifest_contract.py
"""
Run-Manifest-Contract v1 (Issue #763).

Kanonischer, versionierter Vertrag für reproduzierbare Runs:
  - RunManifest: vollständiger Snapshot aller Run-Parameter
  - ReplayRequest: Override-Parameter für Varianten-Replay
  - ReplayResponse: Bestätigung eines gestarteten Replays

Regeln:
  - Keine Secrets im Manifest (API-Keys, Passwörter).
  - Prompt-Texte sind byte-genaue Snapshots zum Zeitpunkt des Runs.
  - Draft-Manifest bei Run-Start, final bei Run-Ende.
  - Legacy-Runs bekommen status="legacy" mit rekonstruierten Feldern.
"""
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, ValidationError, model_validator

from .ai_provider_contract import AiModelRef, AiRoute, RouteSource, _LEGACY_ROUTE_OPTIONS_KEY
from .llm_routing_contract import ReasoningEffort, StageId

_STRICT = ConfigDict(extra="forbid")

ManifestStatus = Literal["draft", "final", "legacy"]


class ManifestInputs(BaseModel):
    """Eingangsdaten-Hashes und -Referenzen.

    ``seed_document_hash``/``seed_document_filename`` sind ``Optional``
    (Issue #1274 Punkt 6): ist die Quelle eines Runs wirklich nicht
    ermittelbar, ist ``None`` ehrlicher als der frühere Platzhalter
    ``"unknown"``, der wie ein echter Wert aussah.
    """

    model_config = _STRICT

    seed_document_hash: Optional[str] = None
    seed_document_filename: Optional[str] = None
    simulation_config_hash: str
    graph_id: str
    graph_version: Optional[str] = None
    embedding_version: Optional[str] = None


class ManifestVersions(BaseModel):
    """Agora- und Schema-Version zum Zeitpunkt des Runs."""

    model_config = _STRICT

    agora_version: str
    schema_version: str


class AiRouteSnapshot(BaseModel):
    """Strikter Manifest-Snapshot einer aufgelösten ``AiRoute`` (Issue #1274 Punkt 2).

    Ersetzt das vorherige offene ``dict[str, Any]``: der interne
    ``__legacy_stage_route__``-Transportkanal von :class:`AiRoute` (siehe
    ``ai_provider_contract.stage_route_from_ai_route``) ist hier bereits
    aufgelöst — ``temperature``/``max_tokens``/``reasoning_effort`` liegen auf
    oberster Ebene — und secret-tragende Provider-Optionen sind entfernt.
    ``extra="forbid"`` lehnt jedes weitere Feld ab, damit kein künftiger
    interner Transportkanal ungeprüft ins Manifest durchsickert.
    """

    model_config = _STRICT

    stage: Optional[StageId] = None
    provider_connection_id: Optional[str] = None
    model_id: Optional[str] = None
    source: RouteSource
    fallback_reason: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    reasoning_effort: ReasoningEffort = "none"
    provider_options: dict[str, Any] = Field(default_factory=dict)


def ai_route_snapshot_from_ai_route(route: AiRoute) -> AiRouteSnapshot:
    """Baut den strikten Manifest-Snapshot aus einer aufgelösten ``AiRoute``.

    Löst den ``__legacy_stage_route__``-Transportkanal auf und entfernt ihn
    zusammen mit ``secret_ref`` aus ``provider_options`` — das Manifest ist
    eine Persistenzgrenze, kein internes Transportformat, und ``secret_ref``
    ist trotz nur einer Referenz-ID kein Feld, das dort weiterreisen muss.
    """
    options = dict(route.provider_options)
    legacy = options.pop(_LEGACY_ROUTE_OPTIONS_KEY, None)
    options.pop("secret_ref", None)
    legacy = legacy if isinstance(legacy, dict) else None
    return AiRouteSnapshot(
        stage=route.stage,
        provider_connection_id=route.provider_connection_id,
        model_id=route.model_id,
        source=route.source,
        fallback_reason=route.fallback_reason,
        temperature=(legacy or {}).get("temperature"),
        max_tokens=(legacy or {}).get("max_tokens"),
        reasoning_effort=(legacy or {}).get("reasoning_effort") or "none",
        provider_options=options,
    )


def _coerce_legacy_ai_route_snapshot(raw: dict[str, Any]) -> Optional[dict[str, Any]]:
    """Alte ``StageRoute.ai_route_snapshot``-Dicts (offener ``AiRoute``-Dump,
    Issue #763..#901) in die strikte :class:`AiRouteSnapshot`-Form übersetzen
    (Issue #1274 Punkt 2).

    Übernimmt bekannte Felder, löst den ``__legacy_stage_route__``-Kanal auf
    und verwirft alles andere (``validated_capabilities``, ``routing_version``,
    ``resolved_at`` existieren im alten ``AiRoute``-Dump, aber nicht im neuen
    Snapshot). Idempotent für bereits neue Snapshots — die kennen keinen
    Legacy-Kanal und tragen bereits nur bekannte Felder. Fehlt ein
    Pflichtfeld (``source``) oder bleibt der normalisierte Snapshot ungültig,
    ist der Altbestand nicht rettbar — ``None`` statt eines kaputten
    Manifests.
    """
    raw_options = raw.get("provider_options")
    options = dict(raw_options) if isinstance(raw_options, dict) else {}
    legacy = options.pop(_LEGACY_ROUTE_OPTIONS_KEY, None)
    legacy = legacy if isinstance(legacy, dict) else None
    options.pop("secret_ref", None)

    normalized: dict[str, Any] = {
        "stage": raw.get("stage"),
        "provider_connection_id": raw.get("provider_connection_id"),
        "model_id": raw.get("model_id"),
        "source": raw.get("source"),
        "fallback_reason": raw.get("fallback_reason"),
        "temperature": (legacy or {}).get("temperature", raw.get("temperature")),
        "max_tokens": (legacy or {}).get("max_tokens", raw.get("max_tokens")),
        "reasoning_effort": (
            (legacy or {}).get("reasoning_effort") or raw.get("reasoning_effort") or "none"
        ),
        "provider_options": options,
    }
    try:
        AiRouteSnapshot.model_validate(normalized)
    except ValidationError:
        return None
    return normalized


class StageRoute(BaseModel):
    """Pro-Stage-Modell- und Provider-Information (ohne Secrets).

    ``base_url`` ist ``Optional`` (Issue #1274 Punkt 4): eine aus
    Run-Metadaten rekonstruierte Legacy-Route kennt oft nur Modell und
    Provider, nie die Basis-URL der Verbindung.
    """

    model_config = _STRICT

    model: str
    provider: str
    base_url: Optional[str] = None
    ai_route_snapshot: Optional[AiRouteSnapshot] = None

    @model_validator(mode="before")
    @classmethod
    def _tolerate_legacy_ai_route_snapshot(cls, data: Any) -> Any:
        """Alte ``ai_route_snapshot``-Dicts vor der Feldvalidierung normalisieren.

        Läuft dem eigentlichen Feld-Parsing von ``ai_route_snapshot`` voran,
        damit ein nicht rettbarer Altbestand als ``None`` statt als
        ``ValidationError`` durch die gesamte Manifest-Lesekette bricht.
        """
        if not isinstance(data, dict):
            return data
        raw = data.get("ai_route_snapshot")
        if raw is None or not isinstance(raw, dict):
            return data
        data = dict(data)
        data["ai_route_snapshot"] = _coerce_legacy_ai_route_snapshot(raw)
        return data


class ManifestRouting(BaseModel):
    """LLM-Routing-Tabelle pro Stage."""

    model_config = _STRICT

    stages: dict[str, StageRoute] = Field(default_factory=dict)


class PromptSnapshot(BaseModel):
    """Byte-genauer Prompt-Text mit Herkunftsangabe."""

    model_config = _STRICT

    content: str
    source_file: str


class ManifestPrompts(BaseModel):
    """Alle Prompt-Snapshots eines Runs."""

    model_config = _STRICT

    entries: dict[str, PromptSnapshot] = Field(default_factory=dict)


class ManifestSeeds(BaseModel):
    """Random-Seed-Informationen.

    ``random_seed`` ist ``Optional`` (Issue #1274 Punkt 1): Agora hat kein
    echtes RNG-Seed-Konzept (kein ``np.random.seed`` o.ä.). Ein aus der
    ``simulation_id`` abgeleiteter Platzhalter-Hash suggerierte hier
    Reproduzierbarkeit, die nicht besteht — ``None`` ist die ehrliche
    Aussage "kein Seed vorhanden". ``simulation_id_seed`` ist ``Optional``
    aus demselben Grund: eine Legacy-Rekonstruktion ohne bekannte
    ``simulation_id`` darf keinen fabrizierten Platzhalter (``"legacy"``)
    tragen.
    """

    model_config = _STRICT

    random_seed: Optional[int] = None
    simulation_id_seed: Optional[str] = None


class ManifestRuntime(BaseModel):
    """Laufzeitdaten — nur im finalen Manifest."""

    model_config = _STRICT

    started_at: AwareDatetime
    completed_at: Optional[AwareDatetime] = None
    duration_seconds: Optional[int] = None
    rounds_completed: Optional[int] = None
    usage_summary: Optional[dict[str, Any]] = None
    termination_reason: Optional[str] = None


class ManifestSimulationParams(BaseModel):
    """Start-Parameter des Simulationslaufs für 1:1-Replay (Issue #1274 Punkt 3).

    Ohne diese Felder setzte ``POST /api/runs/<id>/replay`` stillschweigend
    ``platform="parallel"`` sowie Runner-Defaults für ``max_rounds`` und das
    Graph-Memory-Update statt das Original zu reproduzieren. ``Optional`` auf
    :class:`RunManifest`-Ebene, damit Manifeste vor dieser Änderung lesbar
    bleiben — ein fehlendes Feld dort bedeutet "unbekannt", nicht "Defaults
    galten".
    """

    model_config = _STRICT

    platform: str
    max_rounds: Optional[int] = None
    enable_graph_memory_update: bool
    memory_update_graph_id: Optional[str] = None


class ManifestDeviation(BaseModel):
    """Eine dokumentierte Abweichung eines Replay-Runs vom Original (Issue #1274 Punkt 3).

    Übersteuerbar ist beim Replay ausschließlich die Modell-Route — jede
    Abweichung hier beschreibt also, wo genau ein Override gegenüber dem
    Original gewirkt hat, statt Replay-Runs stillschweigend als 1:1-Kopien
    auszugeben.
    """

    model_config = _STRICT

    field: str
    original: Any = None
    replay: Any = None


class RunManifest(BaseModel):
    """Kanonisches, maschinenlesbares Manifest eines Runs.

    Enthält alle Parameter, die zur Reproduktion nötig sind:
    Eingangsdaten, Versionen, Routing, Prompts, Seeds und Laufzeitdaten.
    """

    model_config = _STRICT

    schema_version: Literal[1] = 1
    run_id: str
    replayed_from_run_id: Optional[str] = None
    captured_at: AwareDatetime

    inputs: ManifestInputs
    versions: ManifestVersions
    routing: ManifestRouting
    prompts: ManifestPrompts
    seeds: ManifestSeeds
    runtime: Optional[ManifestRuntime] = None
    simulation: Optional[ManifestSimulationParams] = None
    deviations: list[ManifestDeviation] = Field(default_factory=list)

    status: ManifestStatus

    @model_validator(mode="after")
    def _final_requires_runtime(self) -> "RunManifest":
        """``status="final"`` ohne Laufzeitdaten verletzt den Lifecycle-Vertrag.

        Ein finales Manifest ist der Reproduktionsanker eines abgeschlossenen
        Runs — ohne ``runtime`` fehlen Start-, Endzeit und Terminierungsgrund.
        ``draft`` (Run läuft noch) und ``legacy`` (Alt-Run ohne erfasste
        Laufzeitdaten) bleiben bewusst ohne ``runtime`` gültig.
        """
        if self.status == "final" and self.runtime is None:
            raise ValueError('status="final" requires runtime data')
        return self


class ReplayOverrides(BaseModel):
    """Override-Parameter für Varianten-Replay.

    Alle Felder optional — nur gesetzte Felder werden überschrieben.
    """

    model_config = _STRICT

    seed_document_id: Optional[str] = None
    random_seed: Optional[int] = None
    # Kanonischer AiModelRef statt offenem dict: erzwingt beide Pflichtfelder
    # (provider_connection_id, model_id) und lehnt unbekannte Schlüssel ab.
    # Gleiche Modell-ID auf zwei Connections ist sonst nicht unterscheidbar.
    ai_model_ref: Optional[AiModelRef] = None


class ReplayRequest(BaseModel):
    """Request-Body für POST /api/runs/<run_id>/replay.

    Leere Overrides = identisches Replay.
    """

    model_config = _STRICT

    overrides: Optional[ReplayOverrides] = None


class ReplayResponse(BaseModel):
    """Response für gestartetes Replay."""

    model_config = _STRICT

    run_id: str
    status: str

File: frontend/src/contracts/runManifestContract.ts
/**
 * Run-Manifest-Contract v1 — Zod-Spiegel (Issue #763).
 *
 * Hand-gepflegt, 1:1 zu backend/app/contracts/run_manifest_contract.py und
 * den generierten JSON-Schemas unter schemas/run-manifest.schema.json etc.
 *
 * Regeln:
 *   - Keine Secrets im Manifest (API-Keys, Passwörter).
 *   - Prompt-Texte sind byte-genaue Snapshots zum Zeitpunkt des Runs.
 *   - Draft-Manifest bei Run-Start, final bei Run-Ende, legacy für Alt-Runs.
 *
 * Änderungen am Pydantic-Modell → Schema-Dump → diese Datei synchronisieren.
 */
import { z } from 'zod'

import { AiModelSourceSchema } from './aiModelRef'
import { RouteSourceSchema } from './aiProviderContract'
import { ReasoningEffortSchema, StageIdSchema } from './llmRoute'

// === ManifestStatus ===
export const ManifestStatusSchema = z.enum(['draft', 'final', 'legacy'])
export type ManifestStatus = z.infer<typeof ManifestStatusSchema>

// === ManifestInputs ===
// seed_document_hash/seed_document_filename sind optional (Issue #1274
// Punkt 6): ist die Quelle eines Runs nicht ermittelbar, ist `null` ehrlicher
// als der frühere Platzhalter "unknown".
export const ManifestInputsSchema = z
  .object({
    seed_document_hash: z.string().nullable().optional(),
    seed_document_filename: z.string().nullable().optional(),
    simulation_config_hash: z.string(),
    graph_id: z.string(),
    graph_version: z.string().nullable().optional(),
    embedding_version: z.string().nullable().optional(),
  })
  .strict()
export type ManifestInputs = z.infer<typeof ManifestInputsSchema>

// === ManifestVersions ===
export const ManifestVersionsSchema = z
  .object({
    agora_version: z.string(),
    schema_version: z.string(),
  })
  .strict()
export type ManifestVersions = z.infer<typeof ManifestVersionsSchema>

// === AiRouteSnapshot (Issue #1274 Punkt 2) ===
// Ersetzt das vorherige offene `z.record(z.string(), z.unknown())`: der
// interne `__legacy_stage_route__`-Transportkanal von AiRoute ist hier
// bereits aufgelöst, secret-tragende Provider-Optionen sind entfernt.
export const AiRouteSnapshotSchema = z
  .object({
    stage: StageIdSchema.nullable().optional(),
    provider_connection_id: z.string().nullable().optional(),
    model_id: z.string().nullable().optional(),
    source: RouteSourceSchema,
    fallback_reason: z.string().nullable().optional(),
    temperature: z.number().nullable().optional(),
    max_tokens: z.number().int().nullable().optional(),
    reasoning_effort: ReasoningEffortSchema.default('none'),
    provider_options: z.record(z.string(), z.unknown()).default(() => ({})),
  })
  .strict()
export type AiRouteSnapshot = z.infer<typeof AiRouteSnapshotSchema>

// === StageRoute ===
// base_url ist optional (Issue #1274 Punkt 4): eine aus Run-Metadaten
// rekonstruierte Legacy-Route kennt oft nur Modell und Provider.
export const StageRouteSchema = z
  .object({
    model: z.string(),
    provider: z.string(),
    base_url: z.string().nullable().optional(),
    ai_route_snapshot: AiRouteSnapshotSchema.nullable().optional(),
  })
  .strict()
export type StageRoute = z.infer<typeof StageRouteSchema>

// === ManifestRouting ===
export const ManifestRoutingSchema = z
  .object({
    stages: z.record(z.string(), StageRouteSchema).default(() => ({})),
  })
  .strict()
export type ManifestRouting = z.infer<typeof ManifestRoutingSchema>

// === PromptSnapshot ===
export const PromptSnapshotSchema = z
  .object({
    content: z.string(),
    source_file: z.string(),
  })
  .strict()
export type PromptSnapshot = z.infer<typeof PromptSnapshotSchema>

// === ManifestPrompts ===
export const ManifestPromptsSchema = z
  .object({
    entries: z.record(z.string(), PromptSnapshotSchema).default(() => ({})),
  })
  .strict()
export type ManifestPrompts = z.infer<typeof ManifestPromptsSchema>

// === ManifestSeeds ===
// random_seed/simulation_id_seed sind optional (Issue #1274 Punkt 1/4): Agora
// hat kein echtes RNG-Seed-Konzept — `null` ist die ehrliche Aussage "kein
// Seed vorhanden" statt eines fabrizierten Platzhalters.
export const ManifestSeedsSchema = z
  .object({
    random_seed: z.number().int().nullable().optional(),
    simulation_id_seed: z.string().nullable().optional(),
  })
  .strict()
export type ManifestSeeds = z.infer<typeof ManifestSeedsSchema>

// === ManifestRuntime ===
export const ManifestRuntimeSchema = z
  .object({
    started_at: z.string(),
    completed_at: z.string().nullable().optional(),
    duration_seconds: z.number().int().nullable().optional(),
    rounds_completed: z.number().int().nullable().optional(),
    usage_summary: z.record(z.string(), z.unknown()).nullable().optional(),
    termination_reason: z.string().nullable().optional(),
  })
  .strict()
export type ManifestRuntime = z.infer<typeof ManifestRuntimeSchema>

// === ManifestSimulationParams (Issue #1274 Punkt 3) ===
// Start-Parameter für 1:1-Replay. Optional auf RunManifest-Ebene, damit
// Manifeste vor dieser Änderung lesbar bleiben — ein fehlendes Feld dort
// bedeutet "unbekannt", nicht "Defaults galten".
export const ManifestSimulationParamsSchema = z
  .object({
    platform: z.string(),
    max_rounds: z.number().int().nullable().optional(),
    enable_graph_memory_update: z.boolean(),
    memory_update_graph_id: z.string().nullable().optional(),
  })
  .strict()
export type ManifestSimulationParams = z.infer<typeof ManifestSimulationParamsSchema>

// === ManifestDeviation (Issue #1274 Punkt 3) ===
export const ManifestDeviationSchema = z
  .object({
    field: z.string(),
    original: z.unknown().nullable().optional(),
    replay: z.unknown().nullable().optional(),
  })
  .strict()
export type ManifestDeviation = z.infer<typeof ManifestDeviationSchema>

// === RunManifest ===
export const RunManifestSchema = z
  .object({
    schema_version: z.literal(1),
    run_id: z.string(),
    replayed_from_run_id: z.string().nullable().optional(),
    captured_at: z.string(),
    inputs: ManifestInputsSchema,
    versions: ManifestVersionsSchema,
    routing: ManifestRoutingSchema,
    prompts: ManifestPromptsSchema,
    seeds: ManifestSeedsSchema,
    runtime: ManifestRuntimeSchema.nullable().optional(),
    simulation: ManifestSimulationParamsSchema.nullable().optional(),
    deviations: z.array(ManifestDeviationSchema).default(() => []),
    status: ManifestStatusSchema,
  })
  .strict()
export type RunManifest = z.infer<typeof RunManifestSchema>

// === ReplayOverrides ===
// ai_model_ref spiegelt den kanonischen Backend-``AiModelRef``: beide
// Pflichtfelder erzwungen, unbekannte Schlüssel abgelehnt. ``source`` hat
// backendseitig einen Default und ist deshalb hier optional.
export const ReplayAiModelRefSchema = z
  .object({
    provider_connection_id: z.string().min(1),
    model_id: z.string().min(1),
    source: AiModelSourceSchema.optional(),
    capability_filter: z.string().nullable().optional(),
    fallback_reason: z.string().nullable().optional(),
  })
  .strict()

export const ReplayOverridesSchema = z
  .object({
    seed_document_id: z.string().nullable().optional(),
    random_seed: z.number().int().nullable().optional(),
    ai_model_ref: ReplayAiModelRefSchema.nullable().optional(),
  })
  .strict()
export type ReplayOverrides = z.infer<typeof ReplayOverridesSchema>

// === ReplayRequest ===
export const ReplayRequestSchema = z
  .object({
    overrides: ReplayOverridesSchema.nullable().optional(),
  })
  .strict()
export type ReplayRequest = z.infer<typeof ReplayRequestSchema>

// === ReplayResponse ===
export const ReplayResponseSchema = z
  .object({
    run_id: z.string(),
    status: z.string(),
  })
  .strict()
export type ReplayResponse = z.infer<typeof ReplayResponseSchema>
